README.md
# commands-support
💾 This EazyAutodelete-Module implements the support of ApplicationCommands.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.4.26](https://github.com/EazyAutodelete/commands-support/compare/v1.4.25...v1.4.26) (2023-07-13)


### Bug Fixes

* **member:** get correct member ([a38a8c1](https://github.com/EazyAutodelete/commands-support/commit/a38a8c178f37f5910ef15f60c310e4d43e292645))

### [1.4.25](https://github.com/EazyAutodelete/commands-support/compare/v1.4.24...v1.4.25) (2023-06-16)


### Bug Fixes

* **errors:** catch errors ([34e8af1](https://github.com/EazyAutodelete/commands-support/commit/34e8af135eee04e723a4b19b7babe404d778cdfc))

### [1.4.24](https://github.com/EazyAutodelete/commands-support/compare/v1.4.23...v1.4.24) (2023-02-03)

### [1.4.23](https://github.com/EazyAutodelete/commands-support/compare/v1.4.22...v1.4.23) (2023-01-08)


### Bug Fixes

* **perms:** use perm names, not numbers ([61ff92d](https://github.com/EazyAutodelete/commands-support/commit/61ff92dda48810a50a5f800be0cbb7b03a8a0bb6))

### [1.4.22](https://github.com/EazyAutodelete/commands-support/compare/v1.4.21...v1.4.22) (2023-01-01)


### Bug Fixes

* allow other component types ([3acb97b](https://github.com/EazyAutodelete/commands-support/commit/3acb97bba30c4fb3ca242c6a57366521a8a08191))

### [1.4.21](https://github.com/EazyAutodelete/commands-support/compare/v1.4.20...v1.4.21) (2022-12-26)


### Bug Fixes

* **selectmenu:** implement new select menu types ([c544e54](https://github.com/EazyAutodelete/commands-support/commit/c544e54c38928a883c78e6fda0ed72e6c1db2dfe))

### [1.4.20](https://github.com/EazyAutodelete/commands-support/compare/v1.4.19...v1.4.20) (2022-12-26)


### Bug Fixes

* **5minlinit:** fix deleting ephemeral messages after 5 mins ([cd008b8](https://github.com/EazyAutodelete/commands-support/commit/cd008b80fc8aa5d3a338173f3ca2a20a7389ae21))

### [1.4.19](https://github.com/EazyAutodelete/commands-support/compare/v1.4.18...v1.4.19) (2022-12-25)

### [1.4.18](https://github.com/EazyAutodelete/commands-support/compare/v1.4.17...v1.4.18) (2022-12-25)


### Bug Fixes

* **disbaled:** let bot mods run disabled cmds ([ae41cb5](https://github.com/EazyAutodelete/commands-support/commit/ae41cb5e9b3b7b37f53c1e147a8db9c1729251a4))

### [1.4.17](https://github.com/EazyAutodelete/commands-support/compare/v1.4.16...v1.4.17) (2022-12-24)


### Bug Fixes

* **cmddisabled:** fix ugly cmd disabled embed ([08aa4df](https://github.com/EazyAutodelete/commands-support/commit/08aa4df00e43b712a853ed2ac882d51382bfe113))

### [1.4.16](https://github.com/EazyAutodelete/commands-support/compare/v1.3.2...v1.4.16) (2022-12-18)


### Bug Fixes

* **core:** use new core ([e40e41f](https://github.com/EazyAutodelete/commands-support/commit/e40e41ffe502a776c301b59a0429970f358bd81f))

### [1.3.2](https://github.com/EazyAutodelete/commands-support/compare/v1.3.0...v1.3.2) (2022-11-19)


### Bug Fixes

* **disabled:** fix translatiuin ([34c3600](https://github.com/EazyAutodelete/commands-support/commit/34c36004c6801ea6c41901b5dcf4f0d5e0db321d))

## [1.3.0](https://github.com/EazyAutodelete/commands-support/compare/v1.2.6...v1.3.0) (2022-11-19)


### Features

* **disabled:** use new commands config ([af79cdd](https://github.com/EazyAutodelete/commands-support/commit/af79cdda1c93609b045245f81465df301cfdb501))

### [1.2.6](https://github.com/EazyAutodelete/commands-support/compare/v1.2.5...v1.2.6) (2022-11-19)

### [1.2.5](https://github.com/EazyAutodelete/commands-support/compare/v1.2.4...v1.2.5) (2022-11-19)

### [1.2.4](https://github.com/EazyAutodelete/commands-support/compare/v1.2.3...v1.2.4) (2022-11-19)

### [1.2.3](https://github.com/EazyAutodelete/commands-support/compare/v1.2.2...v1.2.3) (2022-11-19)

### [1.2.2](https://github.com/EazyAutodelete/commands-support/compare/v1.2.1...v1.2.2) (2022-11-19)

### [1.2.1](https://github.com/EazyAutodelete/commands-support/compare/v1.2.0...v1.2.1) (2022-11-15)


### Bug Fixes

* **disbaledcmd:** fix message for disabled command ([dacd65f](https://github.com/EazyAutodelete/commands-support/commit/dacd65f19929bfd95d5b26ce2d9a0d8aaaf3ca6d))

## 1.2.0 (2022-11-14)


### Features

* add commands support ([4d6ce43](https://github.com/EazyAutodelete/commands-support/commit/4d6ce4396b520b65b838cde392d93a9e6b52b5cc))
* **cmds:** add disabledCommands ([01d04e3](https://github.com/EazyAutodelete/commands-support/commit/01d04e3b0e2f602d281b096a8b4bc987b827e772))
* ignore bot staff on cooldown ([e9e9b82](https://github.com/EazyAutodelete/commands-support/commit/e9e9b82555bd55efb610efa9955eb8665fd82931))


src/CommandSupport.ts
import {
  Bot,
  CommandButtonArgs,
  CommandMessage,
  CommandMessageArgs,
  Command,
  CommandButton,
  Module,
  CommandMenu,
  CommandMenuArgs,
  CommandModal,
  CommandModalArgs,
} from "@eazyautodelete/core";
import { msToDuration, snowflakeToDate } from "@eazyautodelete/bot-utils";
import {
  AutocompleteInteraction,
  CommandInteraction,
  ComponentInteraction,
  GuildChannel,
  Message,
  ModalSubmitInteraction,
  Permission,
} from "eris";

class CommandSupport extends Module {
  disabledCommands: Map<string, string>;
  constructor(bot: Bot) {
    super(bot);
    this.name = "commands-support";

    this.disabledCommands = new Map();
  }

  clientReady() {
    this.logger.info("[💬] CommandSupport available!");
  }

  async interactionCreate(
    interaction: CommandInteraction | AutocompleteInteraction | ComponentInteraction | ModalSubmitInteraction
  ) {
    if (interaction.type === 2) {
      const startAt = performance.now();

      if (!interaction.channel || interaction.channel.type === 1 || !interaction.guildID) {
        const userData = interaction.user?.id ? await this.db.getUserSettings(interaction.user.id) : { language: "en" };
        return await interaction
          .createMessage({
            embeds: [
              {
                color: this.bot.utils.getColor("error"),
                title: ":x: Not supported!",
                description: this.bot.translate("noDMs", userData.language),
              },
            ],
            components: [
              {
                type: 1,
                components: [
                  { type: 2, style: 5, label: "Add EazyAutodelete", url: "https://eazyautodelete.xyz/invite/" },
                ],
              },
            ],
          })
          .catch(this.logger.error);
      }

      const message = new CommandMessage(this.bot, interaction);
      const args = new CommandMessageArgs(message);

      const commandName = args.getCommand();
      const command = <Command>this.bot.commands.get(commandName);
      if (!command) return this.logger.warn(`Command ${commandName} not found!`, "CMD");

      const guild = this.bot.client.guilds.get(interaction.guildID);
      if (!guild) return;

      const member = interaction.member || (interaction.user?.id && guild.members.get(interaction.user.id));
      if (!member) return;

      interaction.member = member;

      await message.loadData();

      const cooldown = this.bot.cooldowns.hasCooldown(commandName, member.user.id);
      if (cooldown) {
        const cooldownEmbed = {
          color: this.bot.utils.getColor("error"),
          description: message.translate(
            "onCooldown",
            msToDuration(cooldown).length > 5 ? msToDuration(cooldown) : "1 second"
          ),
          footer: {
            text: "EazyAutodelete",
            iconURL: this.client.user.avatarURL,
          },
          timestamp: new Date(),
        };

        return await interaction.createMessage({
          embeds: [cooldownEmbed],
          flags: 64,
        });
      }

      if (this.bot.config.commands.disabled.find((x: { name: string; reason: string }) => x.name === commandName)) {
        const disabledReason = this.bot.config.commands.disabled.find(
          (x: { name: string; reason: string }) => x.name === commandName
        ).reason;
        if (disabledReason && !this.bot.permissions.isBotMod(member.user.id)) {
          const commandDisabledEmbed = {
            timestamp: new Date(),
            color: this.bot.utils.getColor("error"),
            description: message.translate("commandDisabled", commandName, disabledReason),
            footer: {
              text: this.client.user.username,
              iconURL: this.client.user.avatarURL,
            },
          };

          return await interaction.createMessage({
            embeds: [commandDisabledEmbed],
            flags: 64,
          });
        }
      }

      const missingBotPerms: string[] = [];
      const channelId = interaction.channel.id;
      const channel = <GuildChannel>(interaction.channel || this.bot.client.getChannel(channelId));
      if (!channel) return;

      const clientMember = guild.members.get(this.client.user.id);
      if (!clientMember) return;

      const botPerms = channel.permissionsOf(this.client.user.id);
      const defaultPerms = ["sendMessages", "embedLinks", "readMessageHistory"];

      defaultPerms.map(s => {
        botPerms.has(s as any) || missingBotPerms.push(s);
      });
      command.botPermissions.map(s => {
        botPerms.has(s) || missingBotPerms.push(s.toLocaleString());
      });
      if (missingBotPerms.length >= 1) {
        const botMissingPermsEmbed = {
          timestamp: new Date(),
          color: this.bot.utils.getColor("error"),
          description: message.translate("missingBotPerms", missingBotPerms.join(", ")),
          footer: {
            text: this.client.user.username,
            iconURL: this.client.user.avatarURL,
          },
        };

        return await interaction.createMessage({
          embeds: [botMissingPermsEmbed],
          flags: 64,
        });
      }

      if (!this.bot.permissions.hasPermsToUseCommand(command, member, message.data.guild)) {
        const noPermsEmbed = {
          timestamp: new Date(),
          color: this.bot.utils.getColor("error"),
          description: message.translate("missingPerms"),
          footer: {
            text: "EazyAutodelete",
            iconURL: this.client.user.avatarURL,
          },
        };

        return await interaction.createMessage({
          embeds: [noPermsEmbed],
          flags: 64,
        });
      }

      try {
        await command.run(message, args);
      } catch (error) {
        this.logger.error(`[CommandsSupport:interactionCreate]: ` + error);
      }

      this.bot.permissions.isBotMod(member.user.id) || this.bot.cooldowns.setCooldown(commandName, member.user.id);

      this.logger.info("Ran command '" + commandName + "' in " + (performance.now() - startAt) + "ms", "CMD");
    } else if (interaction.type === 4) {
      const commandName = interaction.data.name;
      const command = <Command>this.bot.commands.get(commandName);
      if (!command) return this.logger.warn(`Command:Autocomplete ${commandName} not found!`, "CMD");

      const opt = interaction.data.options;
      if (opt[0].type === 3) {
        const autocompleteQuery = opt[0].value;
        if (typeof autocompleteQuery != "string") return;
        const queryResults = await command.autocompleteHandler(autocompleteQuery);

        return await interaction
          .result(queryResults)
          .catch(e => this.logger.error("[CommandsSupport:Autocomplete]: " + e));
      }
    } else if (
      interaction.type === 3 &&
      ((<any>interaction.data).component_type === 3 ||
        (<any>interaction.data).component_type === 5 ||
        (<any>interaction.data).component_type === 6 ||
        (<any>interaction.data).component_type === 7 ||
        (<any>interaction.data).component_type === 8)
    ) {
      const created = snowflakeToDate(interaction.message.id);
      const dur = (new Date().getTime() - created.getTime()) / 1000;

      if (dur >= 300) {
        await interaction.createMessage({
          flags: 64,
          embeds: [
            {
              color: this.bot.utils.getColor("error"),
              description: this.bot.translate(
                "expiredSelect",
                interaction.user ? (await this.bot.db.getUserSettings(interaction.user.id)).language : "en"
              ),
            },
          ],
        });

        if (isMessageDeletable(interaction.message)) await interaction.deleteMessage(interaction.message.id);
        return;
      }

      const menu = new CommandMenu(this.bot, interaction);
      const args = new CommandMenuArgs(menu);
      if (args.isCommand()) {
        const commandName = args.getCommand();
        const command = <Command>this.bot.commands.get(commandName);
        if (!command) return this.logger.warn(`Command:SelectMenu "${commandName}" not found!`, "CMD");

        await menu.loadData();

        return command.selectMenuHandler(menu, args);
      }
    } else if (interaction.type === 3 && interaction.data.component_type === 2) {
      const created = snowflakeToDate(interaction.message.id);
      const dur = (new Date().getTime() - created.getTime()) / 1000;

      if (dur >= 300) {
        await interaction.createMessage({
          flags: 64,
          embeds: [
            {
              color: this.bot.utils.getColor("error"),
              description: this.bot.translate(
                "expiredButton",
                interaction.user ? (await this.bot.db.getUserSettings(interaction.user.id)).language : "en"
              ),
            },
          ],
        });
        if (isMessageDeletable(interaction.message)) await interaction.deleteMessage(interaction.message.id);
        return;
      }

      const button = new CommandButton(this.bot, interaction);
      const args = new CommandButtonArgs(button);

      if (args.isCommand()) {
        const commandName = args.getCommand();
        const command = <Command>this.bot.commands.get(commandName);
        if (!command) return this.logger.warn(`Command:Button ${commandName} not found!`, "CMD");

        await button.loadData();

        return await command.buttonHandler(button, args);
      }
    } else if (interaction.type === 5) {
      const modal = new CommandModal(this.bot, interaction);
      const args = new CommandModalArgs(modal);

      if (args.isCommand()) {
        const commandName = args.getCommand();
        const command = <Command>this.bot.commands.get(commandName);
        if (!command) return this.logger.warn(`Command:Modal ${commandName} not found!`, "CMD");

        await modal.loadData();

        return await command.modalHandler(modal, args);
      }
    }
  }
}

export default CommandSupport;

function isMessageDeletable(message: Message) {
  if (!message.guildID) return message.author.id === message.channel.client.user.id;
  const channel = message.channel;
  if (channel instanceof GuildChannel) {
    const permissions = channel.permissionsOf(channel.client.user.id);
    if (!permissions) return false;
    if (permissions.has("administrator")) return true;
    const timeoutUntil = channel.guild.members.get(channel.client.user.id)?.communicationDisabledUntil;
    return (
      (timeoutUntil && timeoutUntil < Date.now()) ||
      message.author.id === channel.client.user.id ||
      permissions.has("manageMessages")
    );
  }
}


src/CommandSupport.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@eazyautodelete/core");
const discord_js_1 = require("discord.js");
const now = performance.now;
class CommandSupport extends core_1.Module {
    constructor(bot) {
        super(bot);
        this.name = "commands-support";
    }
    clientReady() {
        this.logger.info("[💬] CommandSupport available!");
    }
    interactionCreate(interaction) {
        return __awaiter(this, void 0, void 0, function* () {
            if (interaction.isCommand() || interaction.isMessageContextMenu()) {
                const startAt = now();
                if (!interaction.channel || interaction.channel.type === "DM" || !interaction.guild) {
                    const noDMsButton = new discord_js_1.MessageActionRow().addComponents(new discord_js_1.MessageButton()
                        .setURL("https://eazyautodelete.xyz/invite/")
                        .setStyle("LINK")
                        .setLabel("Add EazyAutodelete"));
                    const noDMsEmbed = new discord_js_1.MessageEmbed()
                        .setColor("#ff0000")
                        .setTitle(":x: Not supported!")
                        .setDescription("Commands via dm are not supported, you need to add EazyAutodelete to a server!");
                    return yield interaction
                        .reply({
                        embeds: [noDMsEmbed],
                        components: [noDMsButton],
                    })
                        .catch(this.logger.error);
                }
                const message = new core_1.CommandMessage(this.bot, interaction);
                const args = new core_1.CommandArgs(message);
                const commandName = args.getCommand();
                const command = this.bot.commands.get(commandName);
                if (!command)
                    return this.logger.warn(`Command ${commandName} not found!`, "CMD");
                const guild = this.client.guilds.cache.get(interaction.guild.id) || (yield this.client.guilds.fetch(interaction.guild.id));
                const member = interaction.member ||
                    guild.members.cache.get(interaction.user.id) ||
                    (yield guild.members.fetch(interaction.user.id).catch(this.logger.error));
                if (!member)
                    return this.logger.error("Failed to load Member " + interaction.user.id + " in Guild " + interaction.guild.id);
                interaction.member = member;
                const cooldown = this.bot.cooldowns.hasCooldown(commandName, member.user.id);
                if (cooldown) {
                    const cooldownEmbed = new discord_js_1.MessageEmbed()
                        .setTimestamp()
                        .setColor(this.bot.utils.getColor("error"))
                        .setDescription(message.translate("onCooldown", cooldown.toString()))
                        .setFooter({
                        text: "EazyAutodelete",
                        iconURL: this.client.user.avatarURL({
                            dynamic: true,
                        }) || "",
                    });
                    return yield interaction.reply({
                        embeds: [cooldownEmbed],
                        ephemeral: true,
                    });
                }
                yield message.loadData();
                // TODO disabledCommands
                /**
                if (client.disabledCommands.has(commandName)) {
                  const disabledReason = client.disabledCommands.get(commandName);
                  if (!disabledReason) return;
          
                  const commandDisabledEmbed = new MessageEmbed()
                    .setTimestamp()
                    .setColor(client.colors.succesfull)
                    .setDescription(message.translate("config.commands.disabled", disabledReason))
                    .setFooter({
                      text: "Questions? => /help",
                      iconURL:
                        client.user?.avatarURL({
                          dynamic: true,
                        }) || undefined,
                    });
          
                  return await interaction.reply({
                    embeds: [commandDisabledEmbed],
                    ephemeral: true,
                  });
                }*/
                const missingBotPerms = [];
                const channel = interaction.channel || (yield guild.channels.fetch(interaction.channelId).catch(this.logger.error));
                if (!channel)
                    return;
                const clientMember = yield guild.members.fetch(this.client.user.id).catch(this.logger.error);
                if (!clientMember)
                    return;
                const botPerms = channel.permissionsFor(clientMember);
                const defaultPerms = [
                    discord_js_1.Permissions.FLAGS.SEND_MESSAGES,
                    discord_js_1.Permissions.FLAGS.EMBED_LINKS,
                    discord_js_1.Permissions.FLAGS.USE_EXTERNAL_EMOJIS,
                ];
                defaultPerms.map(s => {
                    botPerms.has(s) || missingBotPerms.push(s);
                });
                command.botPermissions.map(s => {
                    botPerms.has(s) || missingBotPerms.push(s);
                });
                if (missingBotPerms.length >= 1) {
                    const botMissingPermsEmbed = new discord_js_1.MessageEmbed()
                        .setTimestamp()
                        .setColor(this.bot.utils.getColor("default"))
                        .setDescription(message.translate("missingBotPerms", channel.id, missingBotPerms.join(", ")))
                        .setFooter({
                        text: "Questions? => /help",
                        iconURL: this.client.user.avatarURL({
                            dynamic: true,
                        }) || "",
                    });
                    return yield interaction.reply({
                        embeds: [botMissingPermsEmbed],
                        ephemeral: true,
                    });
                }
                if (!this.bot.permissions.hasPermsToUseCommand(command, member, message.data.guild)) {
                    const noPermsEmbed = new discord_js_1.MessageEmbed()
                        .setTimestamp()
                        .setColor(this.bot.utils.getColor("error"))
                        .setDescription(message.translate("missingPerms"))
                        .setFooter({
                        text: "EazyAutodelete",
                        iconURL: this.client.user.avatarURL({
                            dynamic: true,
                        }) || "",
                    });
                    return yield interaction.reply({
                        embeds: [noPermsEmbed],
                        ephemeral: true,
                    });
                }
                yield command.run(message, args);
                this.bot.permissions.isBotMod(member.user.id) || this.bot.cooldowns.setCooldown(commandName, member.user.id);
                this.logger.info("Ran command '" + commandName + "' in " + (now() - startAt) + "ms", "CMD");
            }
            else if (interaction.isAutocomplete()) {
                const commandName = interaction.commandName;
                const command = this.bot.commands.get(commandName);
                if (!command)
                    return this.logger.warn(`Command:Autocomplete ${commandName} not found!`, "CMD");
                const autocompleteQuery = interaction.options.data[0].value;
                if (typeof autocompleteQuery != "string")
                    return;
                const queryResults = yield command.autocompleteHandler(autocompleteQuery);
                return yield interaction.respond(queryResults);
            }
            else if (interaction.isSelectMenu()) {
                const created = discord_js_1.SnowflakeUtil.deconstruct(interaction.message.id).timestamp;
                const dur = (new Date().getTime() - created) / 1000;
                if (dur >= 300) {
                    yield interaction.reply({
                        ephemeral: true,
                        content: "⏰ You can not use buttons older than 5 minutes.",
                    });
                    if (interaction.message instanceof discord_js_1.Message && interaction.message.deletable)
                        yield interaction.message.delete().catch(this.logger.error);
                    return;
                }
                const menu = new core_1.CommandMenu(this.bot, interaction);
                const args = new core_1.CommandMenuArgs(menu);
                if (args.isCommand()) {
                    const commandName = args.getCommand();
                    const command = this.bot.commands.get(commandName);
                    if (!command)
                        return this.logger.warn(`Command:SelectMenu ${commandName} not found!`, "CMD");
                    yield menu.loadData();
                    return command.selectMenuHandler(menu, args);
                }
            }
            else if (interaction.isButton()) {
                const created = discord_js_1.SnowflakeUtil.deconstruct(interaction.message.id).timestamp;
                const dur = (new Date().getTime() - created) / 1000;
                if (dur >= 300) {
                    yield interaction.reply({
                        ephemeral: true,
                        content: "⏰ You can not use buttons older than 5 minutes.",
                    });
                    if (interaction.message instanceof discord_js_1.Message &&
                        interaction.message.deletable &&
                        interaction.message.flags.bitfield === 0)
                        yield interaction.message.delete().catch(this.logger.error);
                    return;
                }
                const button = new core_1.CommandButton(this.bot, interaction);
                const args = new core_1.CommandButtonArgs(button);
                if (args.isCommand()) {
                    const commandName = args.getCommand();
                    const command = this.bot.commands.get(commandName);
                    if (!command)
                        return this.logger.warn(`Command:Button ${commandName} not found!`, "CMD");
                    yield button.loadData();
                    return yield command.buttonHandler(button, args);
                }
            }
            else if (interaction.isModalSubmit()) {
                const modal = new core_1.CommandModal(this.bot, interaction);
                const args = new core_1.CommandModalArgs(modal);
                if (args.isCommand()) {
                    const commandName = args.getCommand();
                    const command = this.bot.commands.get(commandName);
                    if (!command)
                        return this.logger.warn(`Command:Modal ${commandName} not found!`, "CMD");
                    yield modal.loadData();
                    return yield command.modalHandler(modal, args);
                }
            }
        });
    }
}
exports.default = CommandSupport;
